### ISSUE TITLE
Migrate stardoc source code from bazelbuild/bazel

### ISSUE THREAD

**Issue by UebelAndre** [bazelbuild/stardoc#69]
I was trying to see what it would take to implement #27 but cannot find the source code. I'm assuming it's buried somewhere in [bazelbuild/bazel](https://github.com/bazelbuild/bazel). If this is the case, is there another issue tracking splitting that source code out into this repo? Also, if this is true, would it also be possible to update the [CONTRIBUTING.md](https://github.com/bazelbuild/stardoc/blob/master/CONTRIBUTING.md) doc to call this out?

**Comment by UebelAndre** [comment]
@c-parsons quick question 😄

**Comment by keith** [comment]
It's here https://github.com/bazelbuild/bazel/tree/master/src/main/java/com/google/devtools/build/skydoc

**Comment by UebelAndre** [comment]
`skydoc`!!!! I should have known 😅 

Thanks @keith do you know if there are any plans to move that source into this repo? Having it in [bazelbuild/bazel](https://github.com/bazelbuild/bazel) makes it harder to contribute to.

**Comment by keith** [comment]
Good question. The maintainers will have to provide context

**Comment by UebelAndre** [comment]
@c-parsons @laurentlb @jin 
Hey guys, might one of you be able to answer this?

**Comment by UebelAndre** [comment]
Hey, pinging this issue again. Would love to see the stardoc code moved into here 😄

**Comment by laurentlb** [comment]
@alandonovan would like to change the implementation. If Bazel (e.g. `bazel query`) can output the information about the bzl files, Stardoc could be much simpler and wouldn't have to do a fake evaluation of the files. Once it's done, the code won't need access to Bazel internals and can live here.

So I don't think it makes sense to move the code right now.

**Comment by UebelAndre** [comment]
Ah cool, I didn't realize there was that interdependence between the two tools. Is there another issue tracking that work effort?

**Comment by UebelAndre** [comment]
Maybe @alandonovan might know?

**Comment by alandonovan** [comment]
Hi UebelAndre, I don't really have a tracking issue for this work, but I can sketch it briefly. The current design of skydoc is an unfortunate maintenance problem as it executes BUILD files in an environment that tries to simulate the loading phase of Bazel without actually running Bazel. This means that every change to the production code of the really quite complex BUILD environment in Bazel requires a corresponding change to the fakes in Stardoc. To reduce that burden, the entire Starlark API of the build language has been factored into an artificial separation of interface and implementation, even though there is only one real implementation, leading to sublime poetry like this:
https://github.com/bazelbuild/bazel/blob/450c7ad65bfcffebf753159ee2b617e557ffb4fd/src/main/java/com/google/devtools/build/lib/starlarkbuildapi/cpp/CcBootstrap.java#L53

To me the obvious solution is: ask Bazel for the information. It already knows how to load and execute a .bzl file. Bazel query supports many ways reporting information about targets. What we need is an output format that reports information about a target's BUILD file, the .bzl file its loads (and so on recursively), and the Starlark constants, functions, rules, providers, and so on, defined in those .bzl files. Bazel already has the internal data structures. Once we have that, Skydoc can be entirely implemented as an external client of blaze. (Or, if you prefer, as a Starlark-defined build rule that invokes genquery then processes its output.) The information provided by this output format would be useful for many other tools too.

At this point I have a very rough prototype. I expect we'll finish it in Q4.

**Comment by UebelAndre** [comment]
@alandonovan Awesome! Thanks for the info and I'm really looking forward to that change 😄 

That answers all of my questions for this issue but I feel it might be nice to change the title to `Migrate stardoc source code from bazelbuild/bazel` for future discoverability. Your thoughts? Otherwise, this can be closed.

**Comment by alandonovan** [comment]
@c-parsons Could you rename this issue? Apparently I don't have write access.

**Comment by UebelAndre** [comment]
@alandonovan I can do it. I just wanted to know if you were OK with that. 

@c-parsons I just renamed the issue.

**Comment by UebelAndre** [comment]
@alandonovan Hey, any updates here?

**Comment by pauldraper** [comment]
> Once we have that, Skydoc can be entirely implemented as an external client of blaze.

!!!!!

That would be lovely. Doing the bzl_library/stardoc dance is quite awkward, even more so if you don't want to impose stardoc requirements on downstream consumers.

**Comment by alandonovan** [comment]
Update: Q1 seems more likely.

**Comment by brandjon** [comment]
Update: See #89. We're targeting Q2 at the moment. (On the plus side, since this blocks significant cleanup of Bazel's codebase, it remains a priority.)

**Comment by brandjon** [comment]
Deduping #24 into this thread, see that one for additional discussion.

> That would be lovely. Doing the bzl_library/stardoc dance is quite awkward,

The bzl_library dance is still required whenever you have tooling (within the build) that needs to consume .bzl files and their transitive dependencies. Maybe documentation won't need it, but some kinds of tests might.

> even more so if you don't want to impose stardoc requirements on downstream consumers.

Under the new approach, .bzl files won't have to satisfy any extra syntactic/logical requirements to be stardoc-able, unlike today where imperfect mocking means you have to be careful how you construct your .bzl code lest you crash Stardoc.

But I take it you're talking about how bzl_library adds a constraint on your upstream? I.e. if you want to support downstream consumers bzl_library-ing their .bzl files, you need to bzl_library your own files. Again, that won't be required for Stardoc in the future, but it still may be a good idea for other reasons.

(I actually even forgot that the revamp we're talking about would remove the requirement to use bzl_library. That's a nice little bonus, thanks for reminding me.)

**Comment by pauldraper** [comment]
> But I take it you're talking about how bzl_library adds a constraint on your upstream? I.e. if you want to support downstream consumers bzl_library-ing their .bzl files, you need to bzl_library your own files.

Correct.

For example, Bazel itself does not make `//tools/build_defs/repo:http.bzl` available via a bzl_library (or even a public file for that matter).

So it's literally impossible to use Stardoc for anything that transitively references http_repo.

It's extremely limiting.

**Comment by UebelAndre** [comment]
> So it's literally impossible to use Stardoc for anything that transitively references http_repo, which is a very common repository rule.

This is not entirely true, you can define a `bzl_library` like the following which contains `@bazel_tools//tools/build_defs/repo:http.bzl`
```python
load("@bazel_skylib//:bzl_library.bzl", "bzl_library")

bzl_library(
    name = "bazel_tools_bzl_srcs",
    srcs = ["@bazel_tools//tools:bzl_srcs"],
)
```

I'm not sure if this is an oversight or not but does mean you can use stardoc with rules that use `@bazel_tools//tools/build_defs/repo:http.bzl`

Other than that, I'm **very** excited about this work being done. Do you know if it's still going to be the case that users should maintain `bzl_library` targets? Or would the canonical use of stardoc not require that? I understand if it's too early to tell.

**Comment by pauldraper** [comment]
That only works if the author has used `export_files()`.

In this case, they had not. (Because file visibility is irrelevant for load()).

> or even a public file for that matter

**Comment by UebelAndre** [comment]
> That only works if the author has used `export_files()`.

This would be a great thing to be able to avoid. If I can load the file, I should be able to generate documentation for **my** files that load it.

ps: I had edited [my comment](https://github.com/bazelbuild/stardoc/issues/69#issuecomment-759119947) a bit but all I was trying to say there was it's possible to use stardoc on `.bzl` files that load `http.bzl` but that my solution is likely an oversight rather than an intentionally exposed set of source files.

**Comment by brandjon** [comment]
> Do you know if it's still going to be the case that users should maintain bzl_library targets?

bzl_library won't be needed for the sake of documentation generation but it still may be needed for other tooling that needs to consume .bzl files along with their transitive loads. For instance, if you have a build workflow that does some kind of linting or testing of .bzl files (aside from Bazel's built-in analysis-time testing framework).

> file visibility is irrelevant for load()

And therein lies the reason for .bzl's privileged status in this regard. If visibility were required for loading, you'd at least guarantee the ability to define bzl_library targets by downstream consumers if not in the declaring package. But that's not possible because visibility is a package-level concept, and we don't want to block bzl loading on package loading.

(We do have notions about using a separate visibility mechanism for bzl files. Design doc to come. But that's unrelated to Stardoc.)

**Comment by pauldraper** [comment]
> and we don't want to block bzl loading on package loading.

Though bzl loading does depend on packages. 

It checks if a BUILD.bazel exists, though apparently doesn't care what's in it.

**Comment by pauldraper** [comment]
What I do know is that **Bazel itself** doesn't even make Skylark files available. One way or another, I have little hope that bzl_library-dependent things will be feasible.

https://github.com/bazelbuild/bazel/issues/12444

**Comment by brandjon** [comment]
> It checks if a BUILD.bazel exists, though apparently doesn't care what's in it.

Yes. In the code, this is the distinction between `PackageFunction` and `ContainingPackageLookupFunction`. The latter is much cheaper since it doesn't have to parse and evaluate a BUILD file, which includes creating the representation of all its targets.

> What I do know is that Bazel itself doesn't even make commonly-used Skylark files available (bazelbuild/bazel#12444). One way or another, I have little hope that bzl_library-dependent things will be feasible.

This is really something we should revisit when we look at bzl visibility. If exports_files is the best practice we've got then we should do it everywhere in `@bazel_tools`. But it's possible bzl visibility may supersede the need for that, e.g. if we allow load()-based visibility to imply permission to also depend on the .bzl as a target.

**Comment by UebelAndre** [comment]
Hey, just checking and wondering what the status of this is.

**Comment by tetromino** [comment]
The status is that it's definitely on our list, but we aren't working on it at the moment.

**Comment by akstr** [comment]
Hi,

I wonder how  it is going with the redesign of Stardoc. According https://github.com/bazelbuild/stardoc/blob/master/docs/future_plans.md,  it would be done in Q2.  Is this still valid?

**Comment by UebelAndre** [comment]
Friendly ping 😅

**Comment by tetromino** [comment]
Pushed back to Q3-ish, sorry :/

**Comment by ecngtng** [comment]
Hi, are there any updates for the redesign of Stardoc?😅

**Comment by aiuto** [comment]
One way to expedite this is to  split the redesign/rebuild process at the rendering boundary.  We could fork the renderer and let that development proceed in stardoc while leaving the one internal to Bazel for our own use.

Using the existing Java sources for it does get ugly quickly:
- the renderer is deeply dependent on  net.starlark.java,  so that would have to be available directly outside Bazel. 
- it is also dependent on other bits of Bazel source, so there is rewriting or .jar exporting to be done there.

But, overall, I think a motivated party could fix this without touching Bazel code in a couple of days.  It's not even that much code (~1000 raw source lines of java), someone could build a fresh renderer in whatever language they want in a week. 

This would unblock community contributions to Stardoc by lowering the barrier to development. It would also be compatible with our concept of emitting the stardoc proto directly from Blaze.  It would set us up to be in a place where there is no more stardoc binary.  Just bazel output and a renderer. At that time, bazel could import the next generation renderer to its third_party.

**Comment by tetromino** [comment]
Half of this (the renderer side) was fixed by https://github.com/bazelbuild/stardoc/commit/dbe010734e94081487fb6c47e27085b59daf77aa

The second half will be fixed by an soon-upcoming PR.
